src/infra/core/src/utils/simple_transfer_protocol.rs

```diff
@@ -247,130 +357,42 @@ impl SimpleTransferProtocol {
-                        message: "Source data file is missing".to_owned(),
-                        source: Some(e.into()),
-                    }
-                    .into()),
-                    Err(GetError::Access(e)) => Err(SyncError::Access(e)),
-                    Err(GetError::Internal(e)) => Err(SyncError::Internal(e)),
-                }?;
-
-                match dst
-                    .as_data_repo()
-                    .insert_stream(
-                        stream,
-                        InsertOpts {
-                            precomputed_hash: if !trust_source_hashes {
-                                None
-                            } else {
-                                Some(&data_slice.physical_hash)
-                            },
-                            expected_hash: Some(&data_slice.physical_hash),
-                            size_hint: Some(data_slice.size as usize),
-                            ..Default::default()
-                        },
-                    )
-                    .await
-                {
-                    Ok(_) => Ok(()),
-                    Err(InsertError::HashMismatch(e)) => Err(CorruptedSourceError {
-                        message: concat!(
-                            "Data file hash declared by the source didn't match ",
-                            "the computed - this may be an indication of hashing ",
-                            "algorithm mismatch or an attempted tampering",
-                        )
-                        .to_owned(),
-                        source: Some(e.into()),
-                    }
-                    .into()),
-                    Err(InsertError::Access(e)) => Err(SyncError::Access(e)),
-                    Err(InsertError::Internal(e)) => Err(SyncError::Internal(e)),
-                }?;
-
-                stats.src.data_slices_read += 1;
-                stats.dst.data_slices_written += 1;
-                stats.src.bytes_read += data_slice.size as usize;
-                stats.dst.bytes_written += data_slice.size as usize;
-                listener.on_status(SyncStage::TransferData, &stats);
+        let mut block_download_tasks = vec![];
+        blocks.iter().rev().for_each(|(_, b)| {
+            if let Some(block_stream) = b.as_data_stream_block() {
+                if let Some(data_slice) = block_stream.event.output_data {
+                    block_download_tasks.push(Box::pin(self.download_block_data(
+                        src,
+                        dst,
+                        data_slice,
+                        trust_source_hashes,
+                        listener.clone(),
+                        stats.clone(),
+                    ))
+                    // Each function return unique future
+                    // cast future to next type to allow storing them in vector
+                        as Pin<Box<dyn Future<Output = Result<(), SyncError>> + Send>>)
```

Comment:
This might be done easier if vectors for data block futures and checkpoint futures would be separate.
I don't think it will hurt any parallelism on a realistically large dataset.

Reply:
It was my idea in the first place. I was using try_join to join both data/checkpoint futures and after controlling their amount in the `try_for_each_concurrently`. I may try to implement it and see how its going

Reply:
I've tried to do it in such a way. Here are brief results: When we have two separate vectors for data and checkpoint, we have to run two separate streams, which hurts a lot and performance decreases decently.

If we try to make the way I was thinking about(with try_join inside each iteration) we have 2 issues that somehow block us: 
1. Not each iteration we have both processes and due to this we have to write not smooth if statements where we will return try_join result or just a single future
2. In stream try_for_each_concurrent we have to write a complex error handler to handle errors correctly

Reply:
All right, let it be. At least, we experimented.